### PR TITLE
Add MPI Flavors (Linux/macOS)

### PR DESCRIPTION
**Pull request by ax3l**
Default is `nompi`.

Reference:
- https://conda-forge.org/docs/maintainer/knowledge_base.html#building-mpi-variants
- https://github.com/conda-forge/amrex-feedstock/pull/2

Allows users to install the following:

- `conda install pyamrex` default, nompi variant
- `conda install pyamrex=*=mpi*` any mpi-compatible implementation
- `conda install pyamrex=*=mpi_mpich*` pick mpich variant Same result as `conda install mpich pyamrex=*=mpi*`
- `conda install pyamrex=23.10=nompi*` explicitly forbid MPI Usually not necessary, as the build number offset causes `nompi` to be preferred unless MPI is explicitly requested.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

**Comment by ax3l** [comment]
@minrk do you understand why Intel MPI (`impi`) shows up in the build environment solution of some of the OpenMPI recipes here?

**Comment by minrk** [comment]
It shouldn't, I think folks are working on adding impi to the matrix, but some mutual exclusivity needs to be fixed. I haven't kept up with that process.

**Comment by ax3l** [comment]
Thanks! I am also confused why `mpi4py` sets no `run_exports` in their feedstock on the MPI flavor. Is that safe?

**Comment by leofang** [comment]
AFAIK nothing is fixed yet: https://github.com/conda-forge/mpi-feedstock/issues/11#issuecomment-1769101283. What's the issue, @ax3l?

cc @dalcinl for vis

**Comment by dalcinl** [comment]
Evidently, my addition of Intel MPI mpi4py builds broke things. I don't understand conda-forge enough, I'm not sure how to move forward to fix things.

**Comment by ax3l** [comment]
Ah, no worries. My knowledge is also as much as this knowledge base article:
- https://conda-forge.org/docs/maintainer/knowledge_base.html#building-mpi-variants

Especially this part seems important for an MPI-only Python package like `mpi4py`, I think:
```yaml
build:
# ...

  # add build string so packages can depend on
  # mpi variants explicitly:
  # `pkg * mpi_mpich_*` for mpich
  # `pkg * mpi_*` for any mpi

  {% set mpi_prefix = "mpi_" + mpi %}
  string: {{ mpi_prefix }}_py{{ py }}h{{ PKG_HASH }}_{{ build }}

  run_exports:
    - {{ name }} * {{ mpi_prefix }}_*

# ...

requirements:
  host:
    - {{ mpi }}
  run:
    - {{ mpi }}
```
for `mpi4py` to have the compiled MPI be only allowed combination during runtime.

**Comment by leofang** [comment]
I can help fix it if it's not urgent. Can't do it today.

**Comment by ax3l** [comment]
Thanks @leofang, I opened a PR for review here: https://github.com/conda-forge/mpi4py-feedstock/pull/66

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,5 +1,14 @@
 {% set name = "pyamrex" %}
 {% set version = "23.10" %}
+{% set build = 2 %}
```
Comment (ax3l) [inline recipe/meta.yaml:3]: Due to #5
```suggestion
{% set build = 3 %}
```

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,5 +1,14 @@
 {% set name = "pyamrex" %}
 {% set version = "23.10" %}
+{% set build = 4 %}
```
Comment (ax3l) [inline recipe/meta.yaml:3]: Because of #8
```suggestion
{% set build = 5 %}
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,5 +1,14 @@
 {% set name = "pyamrex" %}
 {% set version = "23.10" %}
+{% set build = 5 %}
```
Comment (ax3l) [inline recipe/meta.yaml:3]: Update after #9:
```suggestion
{% set build = 6 %}
```

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

**Pull request merged**